Smooth Swing Accent changes

SA-22C-smoothswing opened the pull request:
I couldn't get a clean pull from my current fork, so made a new one.  white space is still a mess on smoothswingv2.h but it's better than I was able to get on my original attempt.

Changes are, in brief

- added two new variables for smoothswing.ini (threshold and volume)
- added one new Data object in smooth_swing_v2.h to play the accent swing.  I gave up trying to pass the sounds up to the hybrid font.  This way works well enough
- I check for the presence of swng files
- select random swng file
- set volume of swing to match config file
- play swing file

It all seems to work well, I've tested on a couple fonts with good results.

profezzorn commented on sound/smooth_swing_v2.h:
```diff
@@ -8,36 +8,53 @@
     }
     swings_ = std::min<size_t>(swingl.files_found(), swingh.files_found());
+    if (swng.files_found() > 0)
+    {
```
Please put the { at the end of the previous line, like the rest of this code does.

profezzorn commented on sound/smooth_swing_v2.h:
```diff
@@ -8,36 +8,53 @@
     swings_ = std::min<size_t>(swingl.files_found(), swingh.files_found());
+    if (swng.files_found() > 0)
+    {
+      aswings_ = size_t(swng.files_found());
```
the size_t cast is not needed.

profezzorn commented on sound/smooth_swing_v2.h:
```diff
@@ -214,13 +260,21 @@ class SmoothSwingV2 : public SaberBasePassThrough {
   Data A;
   Data B;
-
+  Data D;
```
We should not use a 'Data' here.
Instead we should use something like:

 RefPtr < BufferedWavPlayer > accent_player_;

instead.

SA-22C-smoothswing replied:
Done.

profezzorn commented on sound/smooth_swing_v2.h:
```diff
@@ -8,36 +8,53 @@
     if (A.player && m - last_random_ < 1000) return;
     last_random_ = m;
-    int swing = random(swings_);
+    while (swing == randomizer)
```
This will hang forever if there is only one swing sound.
Also, we shouldn't pick the sound, we should just play the effect and let it select which one to play.

SA-22C-smoothswing replied:
right, one swing would cause an issue there.  What I was trying to do was prevent the same swing being played twice.

profezzorn commented on sound/smooth_swing_v2.h:
```diff
@@ -94,94 +111,123 @@ class SmoothSwingV2 : public SaberBasePassThrough {
         
       case SwingState::ON:
+        //check for AccentSwingThreshold, presence of accent swings and if the accent player is stopped (this prevents clipping)
+        if (speed >=smooth_swing_config.AccentSwingSpeedThreshold && accent_swings_present && !D.isPlaying() && (A.isPlaying() || B.isPlaying()))
```
I don't think we need to check if A or B is playing since we're in SwingState::ON

SA-22C-smoothswing replied:
I thought so to, but I had accent swings playing while the saber was off until I added that check...

SA-22C-smoothswing replied:
I'm going to leave this check in for safety.  I don't think it's hurting anything and I don't get the accent swings when the saber is off anymore.

profezzorn commented on sound/smooth_swing_v2.h:
```diff
@@ -94,94 +111,123 @@ class SmoothSwingV2 : public SaberBasePassThrough {
         }
         PickRandomSwing();
         state_ = SwingState::OFF;
+        accent_played = false;
```
This is set here, but never used?

SA-22C-smoothswing replied:
leftover from old version, should be removed.

profezzorn commented on sound/smooth_swing_v2.h:
```diff
@@ -94,94 +111,123 @@ class SmoothSwingV2 : public SaberBasePassThrough {
+           accentswing = random(aswings_);
+          }
+          accentrandomizer = accentswing;
+          STDOUT.print("Accent SWING!");
```
Please remove.

profezzorn commented on sound/smooth_swing_v2.h:
```diff
@@ -94,94 +111,123 @@ class SmoothSwingV2 : public SaberBasePassThrough {
+          }
+          accentrandomizer = accentswing;
+          STDOUT.print("Accent SWING!");
+          D.set_volume(smooth_swing_config.AccentSwingVolume);
```
I think we need to make the volume of the accent dynamic, and change as the swing progresses.
If the swing suddenly stops, the volume of the accent swing should also go down to zero I think.

SA-22C-smoothswing replied:
That's a good idea, not quite sure right now how to accomplish that but I'll dig in.

SA-22C-smoothswing replied:
Done.  changed up the config variables.
AccentSwingSpeedThreshold - same, deg/sec needed to trigger accent
AccentSwingVolumeSharpness - new, same idea as SwingSharpness.  Used to govern the fade in/out of the swing sound
MaxAccentSwingVolume - renamed from AccentSwingVolume, combines with AccentSwingVolumeSharpness to govern accent swing volume and fade in/out

SA-22C-smoothswing commented:
ok, I've made all the requested changes and it works really well.  Hopefully these changes meet with your approval as well.  Thanks.

boikonur commented on lightsaber.ino:
```diff
@@ -183,7 +183,7 @@ SnoozeTouch snooze_touch;
 #endif
 
-const char version[] = "$Id$";
+const char version[] = "$Id: 3210a1ba122c79758d6fd90b84af5f57d46b1d2b $";
```
If regular git commiting is used, it wont generate that Id

adays1234 commented:
I notice a few of us are running into whitespace changes.

Clearly you are using whitespace as spaces and 2-char indent levels. Are there other things you are doing for a formatting? Are you using an IDE with a formatter / code cleanup type template? If so, can you post that Prof? Maybe add it to the repo? That way we can all be using the same formatter if we are using IDEs

profezzorn commented:
> I notice a few of us are running into whitespace changes.
> 
> Clearly you are using whitespace as spaces and 2-char indent levels. Are there other things you are doing for a formatting? Are you using an IDE with a formatter / code cleanup type template? If so, can you post that Prof? Maybe add it to the repo? That way we can all be using the same formatter if we are using IDEs

I'm not using any form of IDE. I use emacs, which does some code formatting, but mostly I do it by hand.
PS: I don't recommend emacs to anyone, it's extremely user hostile.
I roughly follow the google style guide for c++: https://google.github.io/styleguide/cppguide.html

profezzorn commented:
I thought about this pull request last night, and I'm wondering if it would be better to let the underlying (wrapped) font play the swing sounds. This has the advantage that it would also work for plecter style swing sounds.

adays1234 commented:
> > I notice a few of us are running into whitespace changes.
> > Clearly you are using whitespace as spaces and 2-char indent levels. Are there other things you are doing for a formatting? Are you using an IDE with a formatter / code cleanup type template? If so, can you post that Prof? Maybe add it to the repo? That way we can all be using the same formatter if we are using IDEs
> 
> I'm not using any form of IDE. I use emacs, which does some code formatting, but mostly I do it by hand.
> PS: I don't recommend emacs to anyone, it's extremely user hostile.
> I roughly follow the google style guide for c++: https://google.github.io/styleguide/cppguide.html

Oh I know emacs. Though I always preferred vi or vim. 

But ok. If you are following the basic cpp guide then a complimentary formatter template is pretty easy. 👍

adays1234 commented:
> I thought about this pull request last night, and I'm wondering if it would be better to let the underlying (wrapped) font play the swing sounds. This has the advantage that it would also work for plecter style swing sounds.

That would simplify things, but would it give the fine-tuned control that sa22c is looking for?

profezzorn commented:
> > I thought about this pull request last night, and I'm wondering if it would be better to let the underlying (wrapped) font play the swing sounds. This has the advantage that it would also work for plecter style swing sounds.
> 
> That would simplify things, but would it give the fine-tuned control that sa22c is looking for?

It certainly could, because there is nothing stopping us from improving/changing/refactoring how hybrid_font.h plays swings.

adays1234 commented:
> > > I thought about this pull request last night, and I'm wondering if it would be better to let the underlying (wrapped) font play the swing sounds. This has the advantage that it would also work for plecter style swing sounds.
> > 
> > 
> > That would simplify things, but would it give the fine-tuned control that sa22c is looking for?
> 
> It certainly could, because there is nothing stopping us from improving/changing/refactoring how hybrid_font.h plays swings.

I actually like that. It simplifies things a lot if smoothswing is simply overlaid into the normal swing players. Presuming that it is configurable to turn on/off and the same tuning can be achieved. It does make the overall structure more simple since it is layers instead of a decision tree in essence.

SA-22C-smoothswing commented on sound/smooth_swing_v2.h:
```diff
@@ -8,39 +8,55 @@
     }
     swings_ = std::min<size_t>(swingl.files_found(), swingh.files_found());
+    //check for swngxx files to use as accent swings
+    if (swng.files_found() > 0){
```
done

SA-22C-smoothswing commented on sound/smooth_swing_v2.h:
```diff
@@ -8,39 +8,55 @@
+      STDOUT.println(swng.files_found());
+      accent_swings_present = true;
+      //allocate player
+      if(!accent_player_){
```
moved to line 121 after the logical check to trigger an accent swing.

SA-22C-smoothswing commented on sound/smooth_swing_v2.h:
```diff
@@ -8,39 +8,55 @@
     swings_ = std::min<size_t>(swingl.files_found(), swingh.files_found());
+    //check for swngxx files to use as accent swings
+    if (swng.files_found() > 0){
+      aswings_ = swng.files_found();
```
we don't if the select function does the randomization for us.

SA-22C-smoothswing commented on sound/smooth_swing_v2.h:
```diff
@@ -8,39 +8,55 @@
+      if(!accent_player_){
+          accent_player_ = GetFreeWavPlayer();
+      }
+    }
```
moved to 121 and resolved.

SA-22C-smoothswing commented on sound/smooth_swing_v2.h:
```diff
@@ -8,39 +8,55 @@
     float start = m / 1000.0;
     A.Stop();
     B.Stop();
+    int swing = random(swings_);
```
removed.

SA-22C-smoothswing commented on sound/smooth_swing_v2.h:
```diff
@@ -214,11 +261,13 @@ class SmoothSwingV2 : public SaberBasePassThrough {
   bool on_ = false;;
   BoxFilter<Vec3, 3> gyro_filter_;
   int swings_;
+  int aswings_;
```
removed

SA-22C-smoothswing commented on sound/smooth_swing_v2.h:
```diff
@@ -94,71 +110,87 @@ class SmoothSwingV2 : public SaberBasePassThrough {
+          accent_player_->PlayOnce(&swng);
+          //select new random swng accent
+          int accentswing = random(aswings_);
+          swng.Select(accentswing);
```
removed

SA-22C-smoothswing commented on sound/smooth_swing_v2.h:
```diff
@@ -167,21 +199,36 @@ class SmoothSwingV2 : public SaberBasePassThrough {
       player->PlayOnce(effect, start);
       player->PlayLoop(effect);
     }
+    void PlayAccent(Effect* effect){
```
removed

SA-22C-smoothswing commented on sound/smooth_swing_v2.h:
```diff
@@ -94,71 +110,87 @@ class SmoothSwingV2 : public SaberBasePassThrough {
-
+          powf(swing_strength, smooth_swing_config.SwingSharpness);
+          
+          float accent_volume =
```
moved volume calculations into if statement to check if accent_player is active.

SA-22C-smoothswing commented on sound/smooth_swing_v2.h:
```diff
@@ -94,71 +110,87 @@ class SmoothSwingV2 : public SaberBasePassThrough {
           float mixhum =
-            powf(swing_strength, smooth_swing_config.SwingSharpness);
-
+          powf(swing_strength, smooth_swing_config.SwingSharpness);
```
fixed.

SA-22C-smoothswing commented:
ok, I've made all the requested changes.

profezzorn commented on lightsaber.ino:
```diff
@@ -540,6 +540,9 @@ public:
     CONFIG_VARIABLE(Transition1Degrees, 45.0f);
     CONFIG_VARIABLE(Transition2Degrees, 160.0f);
     CONFIG_VARIABLE(MaxSwingVolume, 3.0f);
+    CONFIG_VARIABLE(AccentSwingSpeedThreshold, 500.0f);
```
I think this needs to default to zero.

SA-22C-smoothswing replied:
set

profezzorn commented on sound/smooth_swing_v2.h:
```diff
@@ -8,28 +8,39 @@
     }
     swings_ = std::min<size_t>(swingl.files_found(), swingh.files_found());
+    //check for swngxx files to use as accent swings
+    if (swng.files_found() > 0) {
```
&& config_.AccentSwingSpeedThreshold > 0.0

SA-22C-smoothswing replied:
added.

profezzorn commented on sound/smooth_swing_v2.h:
```diff
@@ -8,28 +8,39 @@
+      STDOUT.print("Accent Swings Detected: ");
+      STDOUT.println(swng.files_found());
+      accent_swings_present = true;
+    }
```
} else {

profezzorn commented on sound/smooth_swing_v2.h:
```diff
@@ -8,28 +8,39 @@
     B.Free();
   }
-
+  
```
please undo

profezzorn commented on sound/smooth_swing_v2.h:
```diff
@@ -8,28 +8,39 @@
     B = C;
   }
-
+  
```
please undo

profezzorn commented on sound/smooth_swing_v2.h:
```diff
@@ -49,15 +60,15 @@ class SmoothSwingV2 : public SaberBasePassThrough {
     A.SetTransition(t1_offset, smooth_swing_config.Transition1Degrees);
     B.SetTransition(t1_offset + 180.0,
-      smooth_swing_config.Transition2Degrees);
+           smooth_swing_config.Transition2Degrees);
```
please undo

profezzorn commented on sound/smooth_swing_v2.h:
```diff
@@ -49,15 +60,15 @@ class SmoothSwingV2 : public SaberBasePassThrough {
+           smooth_swing_config.Transition2Degrees);
   }
-
+  
```
please undo

profezzorn commented on sound/smooth_swing_v2.h:
```diff
@@ -49,15 +60,15 @@ class SmoothSwingV2 : public SaberBasePassThrough {
     delegate_->SB_On();
     PickRandomSwing();
-    if (!A.player || !B.player) {
+    if (!A.player || !B.player ) {
```
please undo

profezzorn commented on sound/smooth_swing_v2.h:
```diff
@@ -67,13 +78,13 @@ class SmoothSwingV2 : public SaberBasePassThrough {
     delegate_->SB_Off();
   }
-
+  
```
please undo

profezzorn commented on sound/smooth_swing_v2.h:
```diff
@@ -67,13 +78,13 @@ class SmoothSwingV2 : public SaberBasePassThrough {
     OUT, // Waiting for sound to fade out
   };
-
+  
```
please undo

profezzorn commented on sound/smooth_swing_v2.h:
```diff
@@ -94,38 +105,66 @@ class SmoothSwingV2 : public SaberBasePassThrough {
         
       case SwingState::ON:
+        //check for AccentSwingThreshold, presence of accent swings and if the accent player is stopped (this prevents clipping)
+        if (speed >=smooth_swing_config.AccentSwingSpeedThreshold
```
Please put the && at the end of the lines, and make sure that the new lines line up under each other.

profezzorn commented on sound/smooth_swing_v2.h:
```diff
@@ -94,38 +105,66 @@ class SmoothSwingV2 : public SaberBasePassThrough {
+        if (speed >=smooth_swing_config.AccentSwingSpeedThreshold
+          && accent_swings_present
+            && (A.player->isPlaying() || B.player->isPlaying())
+        {
```
move { to previous line

profezzorn commented on sound/smooth_swing_v2.h:
```diff
@@ -94,38 +105,66 @@ class SmoothSwingV2 : public SaberBasePassThrough {
+            && (A.player->isPlaying() || B.player->isPlaying())
+        {
+          //allocate player
+          if(!accent_player_) {
```
space after if

profezzorn commented on sound/smooth_swing_v2.h:
```diff
@@ -94,38 +105,66 @@ class SmoothSwingV2 : public SaberBasePassThrough {
+          if(!accent_player_) {
+            accent_player_ = GetFreeWavPlayer();
+          }
+          else {
```
This shouldn't be an else.
Instead we need to check if (accent_player_) {
As it is, the first time we get here, we'll allocate the accent player, but not start playing.

SA-22C-smoothswing replied:
ok, resolved.

profezzorn commented on sound/smooth_swing_v2.h:
```diff
@@ -94,38 +105,66 @@ class SmoothSwingV2 : public SaberBasePassThrough {
         if (speed >= smooth_swing_config.SwingStrengthThreshold * 0.9) {
           float swing_strength =
-            std::min<float>(1.0, speed / smooth_swing_config.SwingSensitivity);
+          std::min<float>(1.0, speed / smooth_swing_config.SwingSensitivity);
```
please undo

profezzorn commented on sound/smooth_swing_v2.h:
```diff
@@ -94,38 +105,66 @@ class SmoothSwingV2 : public SaberBasePassThrough {
           while (A.end() < 0.0) {
-            B.midpoint = A.midpoint + 180.0;
-	    Swap();
+           B.midpoint = A.midpoint + 180.0;
```
please undo indentation changes here and below

profezzorn commented on sound/smooth_swing_v2.h:
```diff
@@ -94,38 +105,66 @@ class SmoothSwingV2 : public SaberBasePassThrough {
+            B.set_volume(mixhum * (1.0 - mixab));
+            //This volume will scale with swing speed but is modulated by AccentSwingVolumeSharpness.
+            if (accent_player_) {
+              accent_volume =
```
need to check if accent player is playing.
If it is not, we don't need to calculate and set the volume.
In fact, if it is not, we should free it.

profezzorn commented on sound/smooth_swing_v2.h:
```diff
@@ -142,41 +181,38 @@ class SmoothSwingV2 : public SaberBasePassThrough {
       if (!player) {
-	player = GetFreeWavPlayer();
-	if (!player) return;
+        player = GetFreeWavPlayer();
```
please undo

SA-22C-smoothswing commented:
Provided that this smooth_swing_v2.h is clean and there are no further functional changes, I'd be happy to take a crack at moving these events into the hybrid_font.h file.

profezzorn commented on smooth_swing_v2.h:
```diff
@@ -0,0 +1,259 @@
+#ifndef SOUND_SMOOTH_SWING_V2_H
```
Seems like this file has the wrong name?

SA-22C-smoothswing replied:
uploaded into the wrong directory

profezzorn commented on sound/smooth_swing_v2.h:
```diff
@@ -8,28 +8,34 @@
 public:
   SmoothSwingV2() : SaberBasePassThrough() {}
-
+  
```
please undo

profezzorn commented on sound/smooth_swing_v2.h:
```diff
@@ -94,15 +97,29 @@ class SmoothSwingV2 : public SaberBasePassThrough {
           if (monitor.ShouldPrint(Monitoring::MonitorSwings)) {
-            STDOUT.print("speed: ");
-            STDOUT.println(speed);
+           STDOUT.print("speed: ");
```
please undo (and next line)

profezzorn commented on sound/smooth_swing_v2.h:
```diff
@@ -94,15 +97,29 @@ class SmoothSwingV2 : public SaberBasePassThrough {
       case SwingState::ON:
+        //check for AccentSwingThreshold, presence of accent swings and if the accent player is stopped (this prevents clipping)
+        if (speed >=smooth_swing_config.AccentSwingSpeedThreshold && 
+        accent_swings_present && 
```
the first character (a) needs to go under the "s" in the previous line.
(same with next line)

profezzorn commented on sound/smooth_swing_v2.h:
```diff
@@ -94,15 +97,29 @@ class SmoothSwingV2 : public SaberBasePassThrough {
+        //check for AccentSwingThreshold, presence of accent swings and if the accent player is stopped (this prevents clipping)
+        if (speed >=smooth_swing_config.AccentSwingSpeedThreshold && 
+        accent_swings_present && 
+        (A.player->isPlaying() || B.player->isPlaying())){
```
add space before {

profezzorn commented on sound/smooth_swing_v2.h:
```diff
@@ -94,15 +97,29 @@ class SmoothSwingV2 : public SaberBasePassThrough {
+          if (!accent_player_) {
+            accent_player_ = GetFreeWavPlayer();
+          }
+          if (accent_player){
```
space before {

profezzorn commented on sound/smooth_swing_v2.h:
```diff
@@ -112,20 +129,36 @@ class SmoothSwingV2 : public SaberBasePassThrough {
           while (A.end() < 0.0) {
             B.midpoint = A.midpoint + 180.0;
-	    Swap();
+          Swap();
```
please undo

profezzorn commented on sound/smooth_swing_v2.h:
```diff
@@ -112,20 +129,36 @@ class SmoothSwingV2 : public SaberBasePassThrough {
+            A.set_volume(mixhum * mixab);
+            B.set_volume(mixhum * (1.0 - mixab));
+            //This volume will scale with swing speed but is modulated by AccentSwingVolumeSharpness.
+            if (accent_player_->isPlaying()) {
```
this will crash if there is currently no accent_player_

SA-22C-smoothswing replied:
wrapped it in an if(accent_player) {} check

profezzorn commented on sound/smooth_swing_v2.h:
```diff
@@ -112,20 +129,36 @@ class SmoothSwingV2 : public SaberBasePassThrough {
+              powf(swing_strength, smooth_swing_config.AccentSwingVolumeSharpness);
+              accent_volume *= smooth_swing_config.MaxAccentSwingVolume;
+              accent_player_->set_volume(accent_volume);
+            }
```
} else {

profezzorn commented on sound/smooth_swing_v2.h:
```diff
@@ -142,41 +175,38 @@ class SmoothSwingV2 : public SaberBasePassThrough {
-	player = GetFreeWavPlayer();
-	if (!player) return;
+        player = GetFreeWavPlayer();
+      if (!player) return;
```
please undo

profezzorn commented on sound/smooth_swing_v2.h:
```diff
@@ -142,41 +175,38 @@ class SmoothSwingV2 : public SaberBasePassThrough {
       if (!player) {
-	player = GetFreeWavPlayer();
-	if (!player) return;
+        player = GetFreeWavPlayer();
```
please undo

profezzorn commented on sound/smooth_swing_v2.h:
```diff
@@ -142,41 +175,38 @@ class SmoothSwingV2 : public SaberBasePassThrough {
     delegate_->SetHumVolume(hum_volume);
   }
-
+  
```
please undo

profezzorn commented:
PS: I highly recommend checking how github shows your changes. (essentially reviewing your own code)

SA-22C-smoothswing commented:
Ok, fixed the crash issue and the formatting, I think.  I also uploaded the smoothswing file to the root rather than sound directory.  That has been fixed too.

SA-22C-smoothswing commented:
Ok, got the swings moved to hybrid_font.h

I'm stuck on how to mange volume fading on monophonic swings.  Because a monophonic sound also includes fading out hums, playing the sound and fading hums back in, I'm not sure how to proceed on getting the swing sound broken out so it can be manipulated by smooth_swing_v2.h to perform the volume fades.

profezzorn commented:
Monophonic swing sounds can't fade out in the same way as polyphonic ones can.
We will still adjust the volume slightly based on the speed, which will add some dynamic feeling.
(This would already be handled by SetHumVolume())

profezzorn commented on sound/hybrid_font.h:
```diff
@@ -120,6 +121,36 @@ class HybridFont : public SaberBase {
     }
   }
+  
+  void CommonSwing(Effect* monophonic, Effect* polyphonic) override {
```
Why pass in the effect here?
The hybrid font already knows what effect to use.
Also, I think "StartSwing()" would be a better name.

SA-22C-smoothswing replied:
did this so I could assign a polyphonic_player object and manipulate volume.
I'll see if I can do that without passing the effects and change the function name

SA-22C-smoothswing replied:
I think I need to pass the effects in, but I'll change the function name.

profezzorn commented on sound/hybrid_font.h:
```diff
@@ -120,6 +121,36 @@ class HybridFont : public SaberBase {
+    if (IsSwingPlaying()) {
+      swing_player_->set_volume(volume);
+    }
+    
```
extra empty line

profezzorn commented on sound/hybrid_font.h:
```diff
@@ -173,11 +204,11 @@ class HybridFont : public SaberBase {
-        if (bgnlock.files_found()) {
-          PlayMonophonic(&bgnlock, &lockup);
-        } else {
+	if (bgnlock.files_found()) {
```
why change indentation here?
(and lots of places below)

SA-22C-smoothswing replied:
still fighting whitespace issues.  Thought I had them all resolved.

SA-22C-smoothswing replied:
tried my best, still getting weird whitespace issues.

profezzorn commented on sound/hybrid_font.h:
```diff
@@ -120,6 +121,36 @@ class HybridFont : public SaberBase {
+    }
+  }
+  
+  void SetSwingVolume(float volume) override {
```
I wonder if it would be better if this function just took the current speed, that way it could (optionally) be used by HybridFont::SB_Motion() as well.

Or maybe a speed and a multiplier?

SA-22C-smoothswing replied:
I'll play with that.

profezzorn commented on sound/smooth_swing_v2.h:
```diff
@@ -103,6 +108,12 @@ class SmoothSwingV2 : public SaberBasePassThrough {
         state_ = SwingState::ON;
         
       case SwingState::ON:
+        //check for AccentSwingThreshold, presence of accent swings and if the accent player is stopped (this prevents clipping)
```
This line is too long.
Also, "clipping" is when the sound is too loud to fit in - 32768 - 32768.
I think you mean "prevents cutting off the sound"
Also, we don't actually check if the accent player is stopped here anymore

SA-22C-smoothswing replied:
ok, will update.

SA-22C-smoothswing replied:
ok, moved all the math to hybrid_font.h.  It's pretty specialized for accent swings, but we could change it for other purposes down the road as required.

profezzorn commented on sound/smooth_swing_v2.h:
```diff
@@ -112,20 +123,34 @@ class SmoothSwingV2 : public SaberBasePassThrough {
+            B.set_volume(mixhum * (1.0 - mixab));
+            //This volume will scale with swing speed but is modulated by AccentSwingVolumeSharpness.
+            if (delegate_->IsSwingPlaying()) {
+              accent_volume =
```
I kind of think we should do these volume calculations in hybrid_font.h if possible.

SA-22C-smoothswing replied:
ok, I can pass in the relevant variables and do it there.

SA-22C-smoothswing replied:
done.

profezzorn commented on sound/smooth_swing_v2.h:
```diff
@@ -142,41 +167,37 @@ class SmoothSwingV2 : public SaberBasePassThrough {
           }
         }
         PickRandomSwing();
+        accent_player_.Free();
```
there are still a lot of whitespace changes in this file

SA-22C-smoothswing replied:
Apologies, tried hand-ending to fix, but still showing up as 'changes' in the diff even though I can't see differences when looking at the files.

profezzorn commented on sound/smooth_swing_v2.h:
```diff
@@ -8,28 +8,36 @@
+      STDOUT.print("Monophonic swings: ");
+      STDOUT.println(swing.files_found());
+      accent_swings_present = true;
+    }else{
```
spaces around else

SA-22C-smoothswing commented:
Ok, fixed whitespaces.  I'm seeing some odd behaviour now that I'm testing more.

It seems that accent swings are only triggering half the time I expect them.  For example, if I swing the saber to the right at a speed that is faster than the threshold, I get an accent.  But if I then reverse the swing to the left, at the same (or greater speed) I don't get an accent swing.  Then if I swing again to the right, the accent triggers.

Any idea why that might be?

The other issue I'm having that I'm not sure how to resolve is detecting spins (this might be related to the swing issues listed above).  In reviewing Thexter's original algorithm compared to this implementation, I don't see a way of detecting spins.

SA-22C-smoothswing commented:
I'm also getting audio glitches (pops, clipping) on accent swings when volume is set low (ie: under 1000)

When volume is set to 1000 and up, I don't get the glitches.  Am I setting the volume for the accent swings incorrectly?

profezzorn commented on common/saber_base.h:
```diff
@@ -68,7 +69,11 @@ class SaberBase {
   // This is really just for sound fonts.
   virtual void SetHumVolume(float volume) {}
-
+  virtual void StartSwing(Effect* monophonic, Effect* polyphonic) {}
```
This function does not need to send in effect pointers.
HybridFont already knows what effects to use to start a swing.

SA-22C-smoothswing replied:
fixed.

profezzorn commented on lightsaber.ino:
```diff
@@ -540,6 +540,10 @@ public:
     CONFIG_VARIABLE(Transition2Degrees, 160.0f);
     CONFIG_VARIABLE(MaxSwingVolume, 3.0f);
+    CONFIG_VARIABLE(AccentSwingSpeedThreshold, 0.0f);
+    CONFIG_VARIABLE(AccentSwingVolumeSharpness, 1.0f);
```
I think we should move these variables (except for AccentSwingSpeedThreshold) into config.ini, which can be used directly from hybrid font. That way they can be used to adjust swings for non-smoothswing fonts as well.

SA-22C-smoothswing replied:
ok, i'll drop 'accent' from the parameter names and move them to config.ini  That way we're adjusting 'swing' parameters in config.ini and only calling 'accent' swings from smoothsw.ini

profezzorn replied:
config.ini is the same file that Igniter uses, so we need to use some sort of naming that makes it clear that these variables are intended for ProffieOS. Simplest way is probably to just use replace "Accent" with "ProffieOS". Hopefully Igniter ignores variables it doesn't recognize.

SA-22C-smoothswing replied:
done, moved variables to config.ini

profezzorn commented on sound/hybrid_font.h:
```diff
@@ -120,7 +122,37 @@ class HybridFont : public SaberBase {
   }
+  void StartSwing(Effect* monophonic, Effect* polyphonic) override {
+    if (polyphonic->files_found()) {
+      swing_player_ = PlayPolyphonic(polyphonic);
```
This can cause some weirdness if swing_player_ is already set.
Not sure what we want to happen in that case exactly.

SA-22C-smoothswing replied:
I did this so, when spinning a saber, the sound plays instantly.  Having a check first caused a delay that made swings not play fast enough when spinning.

profezzorn replied:
Oh, I'm not saying that we should not start a swing if one is already playing.
The problem is simply that first swing sound will suddenly not have it's volume adjusted anymore.
I'm not sure if that's a problem, or not, but I thought it would be worth mentioning at least.
If we want to fix it, we would need to possibly keep track of more than one playing swing and adjust all their volumes in SetSwingVolume().

SA-22C-smoothswing replied:
fixed.  Added a second swing player to handle the case where two swings might play.

profezzorn commented on sound/smooth_swing_v2.h:
```diff
@@ -148,6 +166,12 @@ class SmoothSwingV2 : public SaberBasePassThrough {
 	    // We need to stop setting the volume when off, or playback may never stop.
 	    A.set_volume(mixhum * mixab);
 	    B.set_volume(mixhum * (1.0 - mixab));
+      if (delegate_->IsSwingPlaying()) {
```
Let's call IsSwingPlaying inside SetSwingVolume instead.
Then this class will not need to call IsSwingPlaying and it can be removed from SaberBase.

SA-22C-smoothswing commented on sound/hybrid_font.h:
```diff
@@ -104,23 +99,51 @@ class HybridFont : public SaberBase {
     }
   }
-
+  void StartSwing(Effect* monophonic, Effect* polyphonic) override {
```
fixed.  changed to not pass in effects.  Assigns swing_player_ to polyphonic effect, plays monophonic effect normally.

SA-22C-smoothswing commented:
Ah, ok. That make sense.

Matthew McGeary
Senior Advisor, Datacenter, IT
matthew.mcgeary@nutrien.com
306.933.8921
www.nutrien.com

________________________________
From: Fredrik Hubinette <notifications@github.com>
Sent: Monday, May 13, 2019 11:42 AM
To: profezzorn/ProffieOS
Cc: Matthew McGeary; Author
Subject: [EXT] Re: [profezzorn/ProffieOS] Smooth Swing Accent changes (#16)

WARNING: This email originated from outside of the organization. Exercise caution when viewing attachments, clicking links, or responding to requests.


@profezzorn commented on this pull request.

________________________________

In lightsaber.ino<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_profezzorn_ProffieOS_pull_16-23discussion-5Fr283458883&d=DwMCaQ&c=zgHsTqBQkGAdNuoPRxUzSQ&r=8boJ5U4bJexvRn8nMjF_eHn4i8OKnxLC20H4-NjkLr0&m=cty0YnFthw5VdpDqN4unGPJnzx7ja8GSirJiuxUfDD0&s=PR4nGp39dvAKOf-6-fF1LTIESwDOOf2kfhPgp2U6I1M&e=>:

> @@ -540,6 +540,10 @@ public:
     CONFIG_VARIABLE(Transition1Degrees, 45.0f);
     CONFIG_VARIABLE(Transition2Degrees, 160.0f);
     CONFIG_VARIABLE(MaxSwingVolume, 3.0f);
+    CONFIG_VARIABLE(AccentSwingSpeedThreshold, 0.0f);
+    CONFIG_VARIABLE(AccentSwingVolumeSharpness, 1.0f);


config.ini is the same file that Igniter uses, so we need to use some sort of naming that makes it clear that these variables are intended for ProffieOS. Simplest way is probably to just use replace "Accent" with "ProffieOS". Hopefully Igniter ignores variables it doesn't recognize.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_profezzorn_ProffieOS_pull_16-3Femail-5Fsource-3Dnotifications-26email-5Ftoken-3DAL5F72NUFWYVWVGPKNDDVG3PVGSAJA5CNFSM4HIIFWMKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOBYOZ3JY-23discussion-5Fr283458883&d=DwMCaQ&c=zgHsTqBQkGAdNuoPRxUzSQ&r=8boJ5U4bJexvRn8nMjF_eHn4i8OKnxLC20H4-NjkLr0&m=cty0YnFthw5VdpDqN4unGPJnzx7ja8GSirJiuxUfDD0&s=J2Y_jWutJW_vKizzcpfafl1OMSrtZvR-WYQTmxKl4aU&e=>, or mute the thread<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AL5F72IHSQ43A2QGWHMK7JTPVGSAJANCNFSM4HIIFWMA&d=DwMCaQ&c=zgHsTqBQkGAdNuoPRxUzSQ&r=8boJ5U4bJexvRn8nMjF_eHn4i8OKnxLC20H4-NjkLr0&m=cty0YnFthw5VdpDqN4unGPJnzx7ja8GSirJiuxUfDD0&s=iv4-qiWwRCzozEtP_lWcYIfX7-TuS0PDtG5NorZ-slk&e=>.

For more information on Nutrien's email policy or to unsubscribe, click here: https://www.nutrien.com/important-notice
Pour plus de renseignements sur la politique de courrier électronique de Nutrien ou pour vous désabonner, cliquez ici: https://www.nutrien.com/avis-important

SA-22C-smoothswing commented:
Oh, that makes sense now. I’ll put in some logic to check and assign a second player.

Matthew McGeary
Senior Advisor, Datacenter, IT
matthew.mcgeary@nutrien.com
306.933.8921
www.nutrien.com

________________________________
From: Fredrik Hubinette <notifications@github.com>
Sent: Monday, May 13, 2019 11:45 AM
To: profezzorn/ProffieOS
Cc: Matthew McGeary; Author
Subject: [EXT] Re: [profezzorn/ProffieOS] Smooth Swing Accent changes (#16)

WARNING: This email originated from outside of the organization. Exercise caution when viewing attachments, clicking links, or responding to requests.


@profezzorn commented on this pull request.

________________________________

In sound/hybrid_font.h<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_profezzorn_ProffieOS_pull_16-23discussion-5Fr283459942&d=DwMCaQ&c=zgHsTqBQkGAdNuoPRxUzSQ&r=8boJ5U4bJexvRn8nMjF_eHn4i8OKnxLC20H4-NjkLr0&m=jOLUxs9zyo4FJKMTV9LbB27coxEhqA9nooK3szv1Ldk&s=MgHIVd9O_rclOhIlp4F3Y12pSOi3H7k8dG_i54x5g2g&e=>:

> @@ -120,7 +122,37 @@ class HybridFont : public SaberBase {
       PlayPolyphonic(effect);
     }
   }
+  void StartSwing(Effect* monophonic, Effect* polyphonic) override {
+    if (polyphonic->files_found()) {
+      swing_player_ = PlayPolyphonic(polyphonic);


Oh, I'm not saying that we should not start a swing if one is already playing.
The problem is simply that first swing sound will suddenly not have it's volume adjusted anymore.
I'm not sure if that's a problem, or not, but I thought it would be worth mentioning at least.
If we want to fix it, we would need to possibly keep track of more than one playing swing and adjust all their volumes in SetSwingVolume().

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_profezzorn_ProffieOS_pull_16-3Femail-5Fsource-3Dnotifications-26email-5Ftoken-3DAL5F72K2F753ASYACPUGEJDPVGSKVA5CNFSM4HIIFWMKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOBYO2FRY-23discussion-5Fr283459942&d=DwMCaQ&c=zgHsTqBQkGAdNuoPRxUzSQ&r=8boJ5U4bJexvRn8nMjF_eHn4i8OKnxLC20H4-NjkLr0&m=jOLUxs9zyo4FJKMTV9LbB27coxEhqA9nooK3szv1Ldk&s=KWyULccQPhLHfK6xMzFPcHMLq5n-SNNsdGG6G2P3pFw&e=>, or mute the thread<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AL5F72LXQK3G3EGE2BROPHDPVGSKVANCNFSM4HIIFWMA&d=DwMCaQ&c=zgHsTqBQkGAdNuoPRxUzSQ&r=8boJ5U4bJexvRn8nMjF_eHn4i8OKnxLC20H4-NjkLr0&m=jOLUxs9zyo4FJKMTV9LbB27coxEhqA9nooK3szv1Ldk&s=lKkqpADRzGa_5p3tP9vbj5H3DbSyFM-9GzSAzw0zZIg&e=>.

For more information on Nutrien's email policy or to unsubscribe, click here: https://www.nutrien.com/important-notice
Pour plus de renseignements sur la politique de courrier électronique de Nutrien ou pour vous désabonner, cliquez ici: https://www.nutrien.com/avis-important

SA-22C-smoothswing commented:
Ok, changes made as requested.

profezzorn commented on sound/hybrid_font.h:
```diff
@@ -62,11 +68,15 @@ class HybridFont : public SaberBase {
 
   RefPtr<BufferedWavPlayer> hum_player_;
   RefPtr<BufferedWavPlayer> next_hum_player_;
+  RefPtr<BufferedWavPlayer> swing_player_;
```
maybe just make an array:

RefPtr<BufferedWavPlayer> swing_players_[2];

?

SA-22C-smoothswing replied:
decided to have the 'second' swing player not persistent.

profezzorn commented on sound/hybrid_font.h:
```diff
@@ -120,7 +130,49 @@ class HybridFont : public SaberBase {
+          return true;
+        }
+      } else {
+        if (!swing_player_) {
```
This is wrong.
We should free the swing_player_ if it is true but not playing.
Same for the second_swing_player_.

SA-22C-smoothswing replied:
fixed.

SA-22C-smoothswing replied:
removed second_swing_player

profezzorn commented on sound/hybrid_font.h:
```diff
@@ -120,7 +130,49 @@ class HybridFont : public SaberBase {
+    }
+  }
+
+  float SetSwingVolume(float swing_strength, float mixhum) {
```
In order for non-smoothswing fonts to use this, hybrid font has to call this function somewhere.
However, doing so will change font behaviors in ways which I'm not sure are always better.
If we think that updating the swing volume is almost always better than not doing it, that's fine, but otherwise we need the ProffieOSSwingVolumeSharpness to default to 0.0 and ProffiOSMaxSwingVolume to default to whatever the swing defaults to normally? (I don't remember if that is 1.0 or 0.5)

SA-22C-smoothswing replied:
Default swing volume is effectively 1.0.  I'll test some mono/poly fonts with volume changing in place and see how it works.

SA-22C-smoothswing replied:
Setting swing volume seems fine in the polyphonic NEC fonts I've tested.

profezzorn commented on sound/hybrid_font.h:
```diff
@@ -120,7 +130,49 @@ class HybridFont : public SaberBase {
+
+  float SetSwingVolume(float swing_strength, float mixhum) {
+    if (IsSwingPlaying()) {
+      float accent_volume = powf(swing_strength, config_.ProffieOSSwingVolumeSharpness) * config_.ProffieOSMaxSwingVolume;
```
ProffieOSMaxSwingVolume seems to be misnamed, as it is not a maximum.

SA-22C-smoothswing replied:
its the maximum multiple for the swing sound over the hum, much like the smooth swing pairs.

profezzorn commented on sound/hybrid_font.h:
```diff
@@ -120,7 +130,49 @@ class HybridFont : public SaberBase {
+      if (second_swing_player_) {
+        second_swing_player_->set_volume(accent_volume);
+      }
+      mixhum = mixhum * config_.ProffieOSMaxSwingDucking;
```
We really need to make the ducking larger when the accent_volume is larger.
Unconditionally ducking by a fixed amount while the swing is playing is not going to sound good.
It might be better to just not duck at all and let the dynamic mixer sort it out, but at least that is configurable right now.

SA-22C-smoothswing replied:
removed ducking, not required.

profezzorn commented on sound/smooth_swing_v2.h:
```diff
@@ -148,6 +166,7 @@ class SmoothSwingV2 : public SaberBasePassThrough {
 	    // We need to stop setting the volume when off, or playback may never stop.
 	    A.set_volume(mixhum * mixab);
 	    B.set_volume(mixhum * (1.0 - mixab));
+      mixhum = delegate_->SetSwingVolume(swing_strength, mixhum);
```
fix indentation?

profezzorn replied:
are you sure you don't want to do this *before* A.set_volume()
That will duck all the hum volumes, not just the "main" hum.

SA-22C-smoothswing replied:
good point, I've moved it.

SA-22C-smoothswing replied:
removed the mix hum setting.  the ducking wasn't required.

SA-22C-smoothswing commented on sound/hybrid_font.h:
```diff
@@ -120,7 +130,49 @@ class HybridFont : public SaberBase {
   }
+  void StartSwing() override {
+    if (!guess_monophonic_) {
+      if (!swing_player_) {
```
added a check to prevent allocating second_swing_player_ if it's already in use.

SA-22C-smoothswing commented on common/saber_base.h:
```diff
@@ -6,6 +6,7 @@
 // has a corresponding SaberBase::Do* function which invokes that function
 // on all active SaberBases.
 class SaberBase;
+class Effect;
```
removed.

SA-22C-smoothswing commented:
Ok, fixed up some things.  

Removed the concept of a persistent second_swing_player_ and inserted a check in the StartSwing function to ensure that overlapping swings don't occur by checking if at least a quarter of the swing sound has played before fading out the current swing and playing a new one.  This actually improves the saber behaviour with polyphonic NEC fonts as those used to false trigger swings concurrently during a swing, but now that doesn't occur.

Added a parameter to config.ini for swing speed threshold, in order to make saber response to swings tuneable.  Default set to 250, which was the previously hardcoded swing speed trigger.

NEC fonts now get volume fade in/out just like accent swings.  No problems with it so far as I can tell in my testing.

Hopefully we're close to fixing all the issues. 👍

profezzorn commented on sound/hybrid_font.h:
```diff
@@ -120,7 +128,37 @@ class HybridFont : public SaberBase {
+        swing_player_ = PlayPolyphonic(&swng);
+      }
+      // avoid overlapping swings unless at least a quarter of the swing has played.  Stops double swing events.
+      if ((swing_player_->length() - swing_player_->pos()) / swing_player_->length() <= 0.75) {
```
I do like this, although I think we should make the constant configurable.
Also, it seems like pos / length >= 0.25 would be simpler and more clear.

SB_motion() has a different ways to avoid double swings, which is that swing starts once speed goes over a constant, but you can't trigger another swing until speed drops below that constant, and then goes above it again.

Although, we probably need a little hysteresis, so that speed has to drop below 200 or something before swinging_ is set to false again.

SA-22C replied:
That’s a good point about the threshold. I’ll play with that. 

I agree about the overlap constant. I’ll make it a config variable.

SA-22C-smoothswing replied:
ProffieOSSwingOverlay defined.  What should the default be?

SA-22C-smoothswing replied:
I've added a check in the SB_Motion function that only sets swinging to false after speed falls below 80% of the threshold.  Still have to test and see if that value works well.  Should that threshold also be a config variable?

SA-22C-smoothswing replied:
added new config variable ProffieOSSwingLowerThreshold, set to .8 by default.  This works great on monophonic fonts to prevent double swings and still be responsive.

profezzorn commented on sound/hybrid_font.h:
```diff
@@ -120,7 +128,37 @@ class HybridFont : public SaberBase {
+      // avoid overlapping swings unless at least a quarter of the swing has played.  Stops double swing events.
+      if ((swing_player_->length() - swing_player_->pos()) / swing_player_->length() <= 0.75) {
+        RefPtr<BufferedWavPlayer> overlap_swing = swing_player_;
+        swing_player_ = PlayPolyphonic(&swng);
```
how about move this out and after the if statement, then it can be combined with line 135 I think.

SA-22C-smoothswing replied:
I see what you mean.  I've removed the redundant swing_player_ assignment and moved the single call to below the if statement regarding playback position/overlay_swing

profezzorn commented on sound/hybrid_font.h:
```diff
@@ -120,7 +128,37 @@ class HybridFont : public SaberBase {
+      if ((swing_player_->length() - swing_player_->pos()) / swing_player_->length() <= 0.75) {
+        RefPtr<BufferedWavPlayer> overlap_swing = swing_player_;
+        swing_player_ = PlayPolyphonic(&swng);
+        overlap_swing->set_fade_time(0.1);
```
I'm not sure this is what we want to do.
Previously, swing sounds were just left to play out by themselves.

SA-22C replied:
This only applies for rapid swings playing over each other, like spins. The idea is to allow rapid accent swing sounds to play over each other without having to worry about each swing sound and setting volume for each. For normal saber swinging, this doesn’t really factor in.

SA-22C-smoothswing replied:
we could dynamically set the fade time to the remaining duration of the swing.  That way all we're really doing is a cross fade.  I'll give that a try.

SA-22C-smoothswing replied:
does overlap_swing->set_fade_time(overlap_swing->length() - overlap_swing->pos()); look good for a dynamic fade time, based on the remaining play time of an effect?

SA-22C-smoothswing replied:
removed fade time

SA-22C-smoothswing replied:
Put fade time back in.  During testing, odd cutoffs occurred without it.

profezzorn commented on sound/hybrid_font.h:
```diff
@@ -158,7 +196,7 @@ class HybridFont : public SaberBase {
       state_ = STATE_HUM_FADE_OUT;
       PlayPolyphonic(&in);
-    }
+    } 
```
extra whitespace after }

profezzorn commented on sound/smooth_swing_v2.h:
```diff
@@ -112,7 +130,7 @@ class SmoothSwingV2 : public SaberBasePassThrough {
           while (A.end() < 0.0) {
             B.midpoint = A.midpoint + 180.0;
-	    Swap();
+            Swap();
```
whitespace change?

profezzorn commented on sound/smooth_swing_v2.h:
```diff
@@ -103,6 +115,12 @@ class SmoothSwingV2 : public SaberBasePassThrough {
         state_ = SwingState::ON;
         
       case SwingState::ON:
+        //trigger accent swing
```
space after //

profezzorn commented on sound/smooth_swing_v2.h:
```diff
@@ -175,8 +194,8 @@ class SmoothSwingV2 : public SaberBasePassThrough {
       if (!player) {
-	player = GetFreeWavPlayer();
-	if (!player) return;
+        player = GetFreeWavPlayer();
```
whitespace change

profezzorn commented on sound/smooth_swing_v2.h:
```diff
@@ -175,8 +194,8 @@ class SmoothSwingV2 : public SaberBasePassThrough {
-	player = GetFreeWavPlayer();
-	if (!player) return;
+        player = GetFreeWavPlayer();
+        if (!player) return;
```
whitespace change

SA-22C-smoothswing commented:
> I think we're getting near the end.
> I know this is has been a long and difficult process, but I'm happy that you're putting up with all my ideas, suggestions and nitpicks.

I started doing this mostly because I haven't done any serious programming since school.  It's been fun to go through reviews and try and re-learn some of this stuff.  I appreciate all the feedback and rigorous adherence to code style.  I've got the whitespaces fixed in smooth_swing_v2.h and I've made some changes to hybrid_font.h as requested.  I'm still thinking about the swing threshold stuff, more to come today.

SA-22C-smoothswing commented:
Added back in a ducking parameter for mixhum.  Some users are encountering volume clipping during accent swings, testing a ducking parameter to see if that fixes it.  The parameter is combined with swing_strength to reduce mixhum by a percentage between 0 and 100.

profezzorn commented on sound/hybrid_font.h:
```diff
@@ -120,7 +132,49 @@ class HybridFont : public SaberBase {
+        // between 0 (no overlap) and 1.0 (full overlap)
+        if (swing_player_->pos() / swing_player_->length() >= config_.ProffieOSSwingOverlap) {
+          RefPtr<BufferedWavPlayer> overlap_swing = swing_player_;
+          swing_player_ = PlayPolyphonic(&swng);
```
If you move this down a few lines (after the Free()) then there is no need for the overlap_swing variable.

SA-22C replied:
Won’t that introduce a delay in switching swing sounds?

profezzorn replied:
No, and I'm not sure why you would think it would.

SA-22C-smoothswing replied:
seems to function the same with the overlay_swing variable removed and no fade.

SA-22C-smoothswing replied:
after deeper testing across multiple fonts, not having the combination of overlay_player_ and fade out causes odd cut off for some swings.  Having this separate player and the fade and stop function makes all overlays sound as natural as possible.

profezzorn replied:
we can still have fade and stop without having a separate variable.

SA-22C-smoothswing replied:
I'll test again with the proper fade time set in SetSwingVolume but for whatever reason if I don't use a separate variable in my testing, I get an odd 'hiccup' on swing starts for some fonts.

profezzorn replied:
If the code as written above gives you a hiccup, but the other code doesn't, then there is a bug somewhere, and we should find and fix it...

SA-22C-smoothswing replied:
I think the issue arises because we're setting volume on the swing_player so if we don't hand off the overlain swing to a different variable, we get odd audio effects because both swings are getting volume set by SetSwingVolume.

SA-22C-smoothswing replied:
hiccup issue resolved with correct fade-time variable set in SetSwingVolume.  a value of 0.02 fixes hum hiccups and seems to eliminate crackles.  Resolved.

SA-22C-smoothswing replied:
fixed, hiccup and crackles seem resolved with setting a short fade_time in SetSwingVolume (20ms)

profezzorn commented on sound/hybrid_font.h:
```diff
@@ -120,7 +132,49 @@ class HybridFont : public SaberBase {
+        if (swing_player_->pos() / swing_player_->length() >= config_.ProffieOSSwingOverlap) {
+          RefPtr<BufferedWavPlayer> overlap_swing = swing_player_;
+          swing_player_ = PlayPolyphonic(&swng);
+          overlap_swing->set_fade_time(overlap_swing->length() - overlap_swing->pos());
```
Is this really needed? I mean, the swing sound has it's own natural fade-out, doesn't it?

SA-22C replied:
I’ll test without and see.

SA-22C-smoothswing replied:
It's required, odd-sounding cutoffs occur without it.

profezzorn commented on sound/hybrid_font.h:
```diff
@@ -120,7 +132,49 @@ class HybridFont : public SaberBase {
+      }
+      else if (!swing_player_) {
+        swing_player_ = PlayPolyphonic(&swng);
+        swing_player_->set_volume_now(0);
```
This is weird, we should not need to do this.
If this is what makes the crackle go away, then I have a theory:
When a new swing starts, it starts at default volume, which may be significantly higher than what we will set it to in SetSwingVolume(), and maybe the rapid changing of the volume is what causes the crackling.

We should probably be doing two things:
1) set the volume to the correct value, as calculated by SetSwingVolume() *before* we start playing the swing.
2) I'm not sure what the default fade time is, but we probably want to set it to something like 20ms or something as that avoids introducing a lot of noise from setting the volume too rapidly.

SA-22C replied:
Ok I’ll test that out.

SA-22C-smoothswing replied:
All I had to do was set fade_time to 2 just before setting volume in SetSwingVolume and all odd pops and crackles went away.

profezzorn commented on sound/hybrid_font.h:
```diff
@@ -120,7 +132,49 @@ class HybridFont : public SaberBase {
+      if (swing_player_->isPlaying()) {
+        float accent_volume = powf(swing_strength, config_.ProffieOSSwingVolumeSharpness) * config_.ProffieOSMaxSwingVolume;
+        swing_player_->set_volume(accent_volume);
+        mixhum = mixhum - mixhum * (config_.ProffieOSSmoothSwingDucking * swing_strength);
```
seems like we should use accent_volume, not swing strength here
I'm still not convinced we need this.

SA-22C-smoothswing replied:
I'll test with accent_volume and see how it works.  I think we want this parameter for a couple reasons:

Testers have reported audio issues in a couple of circumstances:
1) the volume crackling that appears to be fixed or at least reduced frequency by setting player volume to 0
2) Additional volume crackling issues on loud or bass heavy accent swings that are 'fixed' by reducing accent swing volume.  This parameter lets users and font makers choose which trade-off to make: they can reduce accent swing volume using the max parameter or they can reduce swing pair volume during swings using the ducking parameter.

SA-22C-smoothswing replied:
using accent_Volume works well, and via testing the ducking parameter is required to reduce peaks below 32k

profezzorn commented on sound/smooth_swing_v2.h:
```diff
@@ -16,6 +16,18 @@ class SmoothSwingV2 : public SaberBasePassThrough {
       STDOUT.println("Warning, swingl and swingh should have the same number of files.");
     }
     swings_ = std::min<size_t>(swingl.files_found(), swingh.files_found());
+    //check for swngxx files to use as accent swings
```
add space after //

profezzorn commented on sound/smooth_swing_v2.h:
```diff
@@ -144,11 +162,14 @@ class SmoothSwingV2 : public SaberBasePassThrough {
-	  }
+          if (on_) {
+            // We need to stop setting the volume when off, or playback may never stop.
+            if (accent_swings_present) {
```
we don't need to check if accent swings are present here.
Calling SetSwingVolume should be safe even if accent swings are not present.

SA-22C-smoothswing replied:
It's safe provided that the defaults are set so that swing pairs are not ducked.  I'll make the required changes.

profezzorn commented on sound/hybrid_font.h:
```diff
@@ -120,7 +134,49 @@ class HybridFont : public SaberBase {
+    if(swing_player_) {
+      if (swing_player_->isPlaying()) {
+        float accent_volume = powf(swing_strength, config_.ProffieOSSwingVolumeSharpness) * config_.ProffieOSMaxSwingVolume;
+        swing_player_->set_fade_time(2);
```
This sets the fade time to 2 seconds, which is really slow.
Basically it means that the volume doesn't change.

SA-22C replied:
Hmm.  Well I misunderstood the units on that variable. ;)

SA-22C-smoothswing replied:
set to 0.020 (20ms)

profezzorn commented on sound/hybrid_font.h:
```diff
@@ -120,7 +134,49 @@ class HybridFont : public SaberBase {
+        swing_player_->set_fade_time(2);
+        swing_player_->set_volume(accent_volume);
+        mixhum = mixhum - mixhum * (config_.ProffieOSSmoothSwingDucking * accent_volume);
+      }
```
} else {

profezzorn commented on sound/hybrid_font.h:
```diff
@@ -282,13 +338,16 @@ class HybridFont : public SaberBase {
+      float swing_strength = std::min<float>(1.0, speed / config_.ProffieOSSwingSpeedThreshold);
+      SetSwingVolume(swing_strength, 1.0);
+    } else if (swinging_ && speed <= config_.ProffieOSSwingSpeedThreshold * 
+               ProffieOSSwingLowerThreshold, config_.ProffieOSSwingLowerThreshold) {
```
the comma here probably doesn't do what you think.
Besides, it's probably better to just do:

speed <= config_.ProffieOSSwingLowerThreshold and default it to ~200

SA-22C-smoothswing replied:
that is simpler, yes.  fixed.
